### PR TITLE
LightmapGI: Reduce warnings and increase probe accuracy

### LINKED ISSUE #82642: Vulkan: Baking LightmapGI results in dozens of "Inconsistency found in triangulation while building BSP" messages
### Godot version

v4.2.dev.custom_build [9c51d22d6]

### System information

Ubuntu with Nvidia 2060

### Issue description

```
scene/3d/lightmap_gi.cpp:559 - Inconsistency found in triangulation while building BSP, probe interpolation quality may degrade a bit.
```

This causes some of the probes to not be linked correctly (or at all) with other probes, which causes unnatural light transition between probes (or of the probe is not connected with any other, it will transition instantly to that probe is a dynamic object pass through it, causing a unnatural result) 

image

### Steps to reproduce

Bake MRP

### Minimal reproduction project

[Lightmap.zip](https://github.com/godotengine/godot/files/12778218/Lightmap.zip)

### PR DESCRIPTION
**Pull request by permelin**
Fixes #82642
## The issue

It's very common to get warnings when baking LightmapGI and it's easy to get tens of thousands of them in a single scene.

If the scene is smaller than about 100x100 meters, these warnings are often harmless and don't have a noticeable effect on the indirect lighting of dynamic objects. But for larger scenes, the effect is very much noticeable.

The warnings and the bad accuracy have three causes:

1. A float precision problem in Plane.
2. A common and valid scenario where a partitioning plane can't be found for the BSP tree.
3. Delaunay3D can create invalid triangulation with overlapping tetrahedra in scenes covering a large area.

I made #90701 a separate PR for the last issue since Delaunay3D can be accessed directly from GDScript so that change isn't strictly isolated to LightmapGI.

## Increased tolerance for Plane::has_point()

When creating a `Plane` from a triangle that spans hundreds of meters, `Plane::has_point()` can give false negatives when applied to the second and third point of the triangle with the default epsilon. And `is_point_over()` can return either true of false in these cases.

This wreaks havoc on BSP partitioning.

I've done tests on MRPs and scenes with randomly placed probes over 1000x1000x100m. The precision errors are always a power of two and the largest one I've been able to produce is 1/(1<<13) so I'm using that for tolerance when calling `has_point()`. It is 12x the current tolerance of 1e-5 but still small enough that it shouldn't matter. False positives with this magnitude should not have any effect.

## A better partitioning fallback

Probe tetrahedra are recursively sorted into a BSP tree. All the faces of the tetrahedra are candidate planes to be used for partitioning. As long as the remaining tetrahedra in a subtree have at least two that share an edge or a face, this method will always find a plane that cleanly separates at least one into each new subtree.

But the final two (or very rarely three) tetrahedra in a branch are often either disjoint or share only a single vertex. In this scenario, it is no longer guaranteed that one of the faces is also a separating plane. This is common cause of warnings.

With this knowledge, we can handle this specific case explicitly in a way that always succeeds (as long as the triangulation from Delaunay3D is valid). 

## Bonus: Smaller BSP tree

When the chosen BSP separating plane intersects tetrahedra, they are not split but instead moved into both the subtrees.

The current scoring system to find the best plane only rewards balance. If a plane has one tetrahedron over it, one under and 20 intersecting then it gets a perfect score because 1/1 is perfect balance. This makes the BSP tree much larger than it needs to be.

Reduz actually had a better solution but he commented it out. I took it and I made it more extreme. The scoring now penalizes planes for the number of tetrahedra they intersect.

"Worst" case, the tree is now 30% smaller. Best case, 97% smaller.

I can go into details about why I'm not worried that this will create unbalanced trees, if someone is interested. I've also verified that the max tree depth on all my test scenes is less than before. I expect that lookup times at runtime in general should "always" be improved. And when there are many probe nodes, bake time is also noticeably reduced.

### CODE REVIEW

**Comment by permelin** [comment]
Maybe I was too modest in naming this PR. To be clear, I expect that this together with #90701 will take care of _all_ warnings under all normal circumstances (at least until someone finds a new way to break Delaunay3D).

File: scene/3d/lightmap_gi.cpp [hunk 1]
```diff
@@ -477,20 +482,59 @@ int32_t LightmapGI::_compute_bsp_tree(const Vector<Vector3> &p_points, const Loc
...
+				}
+				Vector3 v1 = p_points[simplex1.vertices[j]];
+				for (uint32_t k = j + 1; k < 4; k++) {
+					if (simplex0.vertices[i] == simplex1.vertices[k]) {
+						break;
+					}
+					Vector3 v2 = p_points[simplex1.vertices[k]];
+
+					Plane plane(v0, v1, v2);
+					if (plane == Plane()) { // When v0, v1, and v2 are collinear, they can't form a plane
```
Comment (AThousandShips) [inline scene/3d/lightmap_gi.cpp:523]:
```suggestion
					if (plane == Plane()) { // When v0, v1, and v2 are collinear, they can't form a plane.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).

File: scene/3d/lightmap_gi.cpp [hunk 2]
```diff
@@ -419,7 +422,9 @@ int LightmapGI::_bsp_get_simplex_side(const Vector<Vector3> &p_points, const Loc
 //#define DEBUG_BSP
 
 int32_t LightmapGI::_compute_bsp_tree(const Vector<Vector3> &p_points, const LocalVector<Plane> &p_planes, LocalVector<int32_t> &planes_tested, const LocalVector<BSPSimplex> &p_simplices, const LocalVector<int32_t> &p_simplex_indices, LocalVector<BSPNode> &bsp_nodes) {
-	//if we reach here, it means there is more than one simplex
+	ERR_FAIL_COND_V(p_simplex_indices.size() == 0, -1);
```
Comment (AThousandShips) [inline scene/3d/lightmap_gi.cpp:425]:
```suggestion
	ERR_FAIL_COND_V(p_simplex_indices.is_empty(), -1);
```

**Review by Calinou — APPROVED** [review]
Tested locally (with #90702 applied on top), it works as expected.

This testing project previously resulted in 200+ warnings being printed on every bake, now it doesn't result in any warning: [test_lightmap_preview_bake_4.x.zip](https://github.com/godotengine/godot/files/14981644/test_lightmap_preview_bake_4.x.zip)[^1]

Code looks good to me at a glance.

[^1]: I intentionally placed multiple probes at the same location in the testing project to see if it breaks, but it appears to bake just fine and dynamic object lighting still behaves correctly.

File: scene/3d/lightmap_gi.cpp [hunk 3]
```diff
@@ -477,20 +482,59 @@ int32_t LightmapGI::_compute_bsp_tree(const Vector<Vector3> &p_points, const Loc
...
+					break;
+				}
+				Vector3 v1 = p_points[simplex1.vertices[j]];
+				for (uint32_t k = j + 1; k < 4; k++) {
+					if (simplex0.vertices[i] == simplex1.vertices[k]) {
+						break;
+					}
+					Vector3 v2 = p_points[simplex1.vertices[k]];
+
+					Plane plane(v0, v1, v2);
```
Comment (Calinou) [inline scene/3d/lightmap_gi.cpp:522]: We tend to prefer this style:

```suggestion
					Plane plane = Plane(v0, v1, v2);
```

File: scene/3d/lightmap_gi.cpp [hunk 4]
```diff
@@ -419,7 +422,9 @@ int LightmapGI::_bsp_get_simplex_side(const Vector<Vector3> &p_points, const Loc
 //#define DEBUG_BSP
 
 int32_t LightmapGI::_compute_bsp_tree(const Vector<Vector3> &p_points, const LocalVector<Plane> &p_planes, LocalVector<int32_t> &planes_tested, const LocalVector<BSPSimplex> &p_simplices, const LocalVector<int32_t> &p_simplex_indices, LocalVector<BSPNode> &bsp_nodes) {
-	//if we reach here, it means there is more than one simplex
+	ERR_FAIL_COND_V(p_simplex_indices.size() == 0, -1);
+	ERR_FAIL_COND_V(p_simplex_indices.size() == 1, -(p_simplex_indices[0] + 1));
```
Comment (AThousandShips) [inline scene/3d/lightmap_gi.cpp:426]: Why does this return a different value? Errors shouldn't be relied upon value wise I'd say

File: scene/3d/lightmap_gi.cpp (hunk 3 above)
Comment (permelin) [reply scene/3d/lightmap_gi.cpp:522]: I'll change it. (I actually copy-pasted this line from another place in the same file.)

File: scene/3d/lightmap_gi.cpp (hunk 4 above)
Comment (permelin) [reply scene/3d/lightmap_gi.cpp:426]: That is a fair question. If you call this function directly with only one simplex, it's an error that can't really be gracefully recovered from. But this function is also called recursively. And while it should never be called with only one simplex like that either, we can actually salvage the situation with a valid return value then.

Comment (AThousandShips) [reply scene/3d/lightmap_gi.cpp:426]: I'd say it should either return `-1` and throw an error, or recover without an error and have a valid return value, not return a valid value *and* error, the assumption would be that error means unusable (and should return a value that implies this when possible)

Comment (permelin) [reply scene/3d/lightmap_gi.cpp:426]: OK

**Review by clayjohn — APPROVED** [review]
looks good to me! Let's go ahead with this

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**